telepresence/outbound/cidr.py

```diff
@@ -0,0 +1,292 @@
+# limitations under the License.
+
+import ipaddress
+import json
+from itertools import chain
+from subprocess import CalledProcessError
+from typing import List
+
+from telepresence.proxy import RemoteInfo
+from telepresence.runner import Runner
+from telepresence.utilities import random_name
+
+
+def covering_cidr(ips: List[str]) -> List[str]:
+    """
+    Given list of IPs, return CIDR that covers them all.
+
+    Presumes it's at least a /24.
+    """
+    def collapse(ns):
+        return list(ipaddress.collapse_addresses(ns))
+
+    assert len(ips) > 0
+    ip_addresses = map(ipaddress.IPv4Address, ips)
+    networks = collapse([
+        ipaddress.IPv4Interface(str(ip) +
+                                ("/31" if ip.is_global else "/24")).network
+        for ip in ip_addresses
+    ])
+    # Increase network size until it combines everything:
+    results = []
+    while len(networks) > 1:
+        network = networks[0]
+        rest = networks[1:]
+        supernet = network.supernet()
+
+        while not supernet.is_global:
+            collapsed_networks = collapse([supernet] + rest)
+            # Keep the supernet if it did collapse networks together.
+            if len(collapsed_networks) <= len(rest):
+                network = supernet
+                rest = [n for n in rest if not n.subnet_of(network)]
+
+            supernet = supernet.supernet()
+
+        networks = rest
+        results.append(network)
+
+    return [n.with_prefixlen for n in chain(results, networks)]
+
+
+# Script to dump resolved IPs to stdout as JSON list:
+
+_GET_IPS_PY = """
+import socket, sys, json
+
+result = []
+for host in sys.argv[1:]:
+    host_ips = []
+    for x in socket.getaddrinfo(host, None):
+         if x[:2] == (socket.AF_INET, socket.SOCK_STREAM):
+            host_ips.append(x[4][0])
+    result.append(host_ips)
+sys.stdout.write(json.dumps(result))
+sys.stdout.flush()
+"""
+
+
+def get_proxy_cidrs(
+    runner: Runner,
+    remote_info: RemoteInfo,
+    hosts_or_ips: List[str] = []
+) -> List[str]:
+    """
+    Figure out which IP ranges to route via sshuttle.
+
+    1. Given the IP address of a service, figure out IP ranges used by
+       Kubernetes services.
+    2. Extract pod ranges from API.
+    3. Any hostnames/IPs given by the user using --also-proxy.
+
+    See https://github.com/kubernetes/kubernetes/issues/25533 for eventual
+    long-term solution for service CIDR.
+    """
+
+    span = runner.span()
+
+    # Run script to convert --also-proxy hostnames to IPs, doing name
+    # resolution inside Kubernetes, so we get cloud-local IP addresses for
+    # cloud resources:
+    result = set(k8s_resolve(runner, remote_info, hosts_or_ips))
+    context_cache = runner.cache.child(runner.kubectl.context)
+    result.update(context_cache.lookup("podCIDRs", lambda: podCIDRs(runner)))
+    result.update(
+        context_cache.lookup("serviceCIDR", lambda: serviceCIDR(runner))
+    )
```

Comment:
Please update the lookup key to "serviceCIDRs" with an "s" at the end.

If you don't do this, then every existing user will get weird crashes from `sshuttle` until the cache entry is cleared. The old cache stored a string, the new one stores a list, and the code doesn't check that it got back what it expected.

Reply:
Right! Done